### PR TITLE
Create user-story-based assemblies for external authentication scenarios

### PR DESCRIPTION
**Pull request by asteflova**
This PR is the first step in an attempt to create user-story-based assemblies for external authentication. When finished, these assemblies should guide users through every step of the configuration journey for each particular scenario: from prerequisites to expected result (i.e. user login procedures for verification).

The changes in this PR include:

* Creating assembly files for each config scenario involving external authentication, such as _Configuring [...] as an external identity provider for Project_.  (The assemblies are based on user stories that we've prepared in collaboration with Eng and QE).
* Filling some of the new assemblies' introductions with wording from the 'custom ToC/decision tree' included in the chapter's introduction.
* Removing that custom ToC which is now obsolete.
* Improving some headings to provide more context

The changes in this PR do not include changing the existing content and wording itself; that will come in future PRs.

For context: We have a request for documentation to clarify what exactly is supported with regards to possible external authentication setups. However, clarifying that is tricky if the documentation does not offer clear step-by-step configuration scenarios. By documenting external authentication setups as user-story-based scenarios, rather than as lists of features, we can steer users towards the configuration setups that our team can actively test and support.

* [x] I am familiar with the [contributing](https://github.com/theforeman/foreman-documentation/blob/master/CONTRIBUTING.md) guidelines.

Please cherry-pick my commits into:

* [ ] Foreman 3.10/Katello 4.12
* [ ] Foreman 3.9/Katello 4.11 (planned Satellite 6.15)
* [ ] Foreman 3.8/Katello 4.10
* [ ] Foreman 3.7/Katello 4.9 (Satellite 6.14)
* [ ] Foreman 3.6/Katello 4.8
* [ ] Foreman 3.5/Katello 4.7 (Satellite 6.13; orcharhino 6.6/6.7)
* [ ] Foreman 3.4/Katello 4.6 (EL8 only)
* [ ] Foreman 3.3/Katello 4.5 on EL7 & EL8 (Satellite 6.12 on EL8 only; orcharhino 6.4/6.5 on EL8 only)
* We do not accept PRs for Foreman older than 3.3.

### CODE REVIEW

File: guides/common/assembly_configuring-keycloak-for-external-authentication-to-project-with-totp-cards.adoc [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+include::modules/proc_configuring-project-settings-for-keycloak-authentication-using-the-cli.adoc[leveloffset=+1]
+
+include::modules/proc_configuring-project-with-keycloak-for-totp-authentication.adoc[leveloffset=+1]
+
+include::modules/proc_logging-in-to-the-project-web-ui-using-keycloak-totp-authentication.adoc[leveloffset=+1]
+
+include::modules/proc_logging-in-to-the-project-CLI-using-keycloak.adoc[leveloffset=+1]
+
+include::modules/proc_configuring-group-mapping-for-keycloak-authentication.adoc[leveloffset=+1]
+
```
Comment (maximiliankolb) [inline guides/common/assembly_configuring-keycloak-for-external-authentication-to-project-with-totp-cards.adoc:20]:
```suggestion
```

File: guides/common/modules/con_configuring-ad-integrated-with-freeipa-through-cross-forest-kerberos-trust-as-an-external-identity-provider-for-project.adoc [hunk 2]
```diff
@@ -0,0 +1,6 @@
+[id="configuring-ad-integrated-with-freeipa-through-cross-forest-kerberos-trust-as-an-external-identity-provider-for-project_{context}"]
+= Configuring Active Directory integrated with {FreeIPA} through cross-forest Kerberos trust as an external identity provider for {Project}
+
+Kerberos can create `cross-forest trust` that defines a relationship between two otherwise separate domain forests.
+A domain forest is a hierarchical structure of domains; both AD and {FreeIPA} constitute a forest.
+With a trust relationship enabled between AD and {FreeIPA}, users of AD can access Linux hosts and services using a single set of credentials.
```
Comment (maximiliankolb) [inline guides/common/modules/con_configuring-ad-integrated-with-freeipa-through-cross-forest-kerberos-trust-as-an-external-identity-provider-for-project.adoc:6]:
```suggestion
With a trust relationship enabled between AD and {FreeIPA}, AD users can access Linux hosts and services using a single set of credentials.
```

File: guides/common/modules/con_configuring-an-ldap-server-as-an-external-identity-provider-for-project.adoc [hunk 3]
```diff
@@ -0,0 +1,9 @@
+[id="configuring-an-ldap-server-as-an-external-identity-provider-for-project_{context}"]
+= Configuring an LDAP server as an external identity provider for {Project}
+
+Lightweight Directory Access Protocol (LDAP) is a set of open protocols used to access centrally stored information over a network. With {Project}, you can manage LDAP entirely through the {ProjectWebUI}.
```
Comment (maximiliankolb) [inline guides/common/modules/con_configuring-an-ldap-server-as-an-external-identity-provider-for-project.adoc:4]:
```suggestion
Lightweight Directory Access Protocol (LDAP) is a set of open protocols used to access centrally stored information over a network.
With {Project}, you can manage LDAP entirely through the {ProjectWebUI}.
```

File: guides/common/modules/con_configuring-an-ldap-server-as-an-external-identity-provider-for-project.adoc [hunk 4]
```diff
@@ -0,0 +1,9 @@
...
+
+[NOTE]
```
Comment (maximiliankolb) [inline guides/common/modules/con_configuring-an-ldap-server-as-an-external-identity-provider-for-project.adoc:6]: Same as above. I lean towards "no", but your call Anet!

File: guides/common/modules/con_configuring-keycloak-as-an-openid-provider-for-project.adoc [hunk 5]
```diff
@@ -1,4 +1,4 @@
 [id="Configuring_Project_with_Keycloak_Authentication_{context}"]
```
Comment (maximiliankolb) [inline guides/common/modules/con_configuring-keycloak-as-an-openid-provider-for-project.adoc:1]: please adjust the anchor.

File: guides/common/modules/con_configuring-keycloak-for-external-authentiation-to-project-with-totp-cards.adoc [hunk 6]
```diff
@@ -0,0 +1,4 @@
+[id="Configuring_Keycloak_Authentication_with_TOTP_{context}"]
```
Comment (maximiliankolb) [inline guides/common/modules/con_configuring-keycloak-for-external-authentiation-to-project-with-totp-cards.adoc:1]: please adjust the anchor.

File: guides/common/modules/con_external-authentication-for-provisioned-hosts.adoc [hunk 7]
```diff
@@ -1,5 +1,8 @@
 [id="External_Authentication_for_Provisioned_Hosts_{context}"]
```
Comment (maximiliankolb) [inline guides/common/modules/con_external-authentication-for-provisioned-hosts.adoc:1]: please adjust the anchor.

File: guides/common/modules/con_external-authentication-for-provisioned-hosts.adoc [hunk 8]
```diff
@@ -1,5 +1,8 @@
...
-= External authentication for provisioned hosts
+= Configuring {Project} to manage the lifecycle of a system registered to a {FreeIPA} realm
```
Comment (maximiliankolb) [inline guides/common/modules/con_external-authentication-for-provisioned-hosts.adoc:2]:
```suggestion
= Configuring {Project} to manage the lifecycle of a host registered to a {FreeIPA} realm
```

title does not match file name.

File: guides/common/modules/con_configuring-external-authentication.adoc [hunk 9]
```diff
@@ -17,40 +17,3 @@ For example, to check if entries for `puppet`, `{apache-user}`, `foreman` and `f
 # cat /etc/passwd | grep 'puppet\|{apache-user}\|foreman\|foreman-proxy'
 # cat /etc/group | grep 'puppet\|{apache-user}\|foreman\|foreman-proxy'
```
Comment (ekohl) [inline guides/common/modules/con_configuring-external-authentication.adoc:18]: Unrelated, but I wonder if this is better:
```suggestion
# grep 'puppet\|{apache-user}\|foreman\|foreman-proxy' /etc/passwd /etc/group
```

**Review by ekohl — COMMENTED** [review]
Does this relate to what we discussed yesterday with @lumarel about the two Keycloak versions?

**Comment by asteflova** [comment]
> Does this relate to what we discussed yesterday with @lumarel about the two Keycloak versions?

I don't think so, but then again, I have no idea who @lumarel is (hello!).

@ekohl If you want to know more about why I'm doing this, I have some d/s resources I could share. And similarly, if you have any information on any work that should be done around the Keycloak scenarios, I'd love to hear more.

**Comment by ekohl** [comment]
> I don't think so, but then again, I have no idea who @lumarel is (hello!).

During https://community.theforeman.org/t/documentation-team-meeting-2024-04-04/37381/2 we discussed https://github.com/theforeman/foreman-documentation/pull/2926. The story is that Red Hat IdM will change URLs depending on whether it's powered by Wildfly (current) or Quarkus (future). That has implications for the end user and we may have additional chapters.

**Comment by asteflova** [comment]
> During https://community.theforeman.org/t/documentation-team-meeting-2024-04-04/37381/2 we discussed #2926. The story is that Red Hat IdM will change URLs depending on whether it's powered by Wildfly (current) or Quarkus (future). That has implications for the end user and we may have additional chapters.

Thanks for the additional details. Looking at #2926, it doesn't seem to be a blocker for the changes I'm making. If anything, with a user-story-based structure, it shouldn't be a problem to have one assembly for each of the scenarios (Wildfly and Quarkus). As always, please let me know if I'm missing something.

**Comment by asteflova** [comment]
I applied @maximiliankolb's feedback (thanks!) and updated headings, IDs, and file names to match each other.

File: guides/common/modules/con_configuring-keycloak-for-external-authentiation-to-project-with-totp-cards.adoc
Comment (maximiliankolb) [inline guides/common/modules/con_configuring-keycloak-for-external-authentiation-to-project-with-totp-cards.adoc:1]: file name, anchor, and title do not match

Comment (asteflova) [reply guides/common/modules/con_configuring-keycloak-for-external-authentiation-to-project-with-totp-cards.adoc:1]: Updated.

**Comment by asteflova** [comment]
@maximiliankolb thanks again, _now_ headings, IDs, and file names match :crossed_fingers:

**Review by maximiliankolb — APPROVED** [review]
Thanks Anet, diff LGTM. I also had a look at the built docs locally and it looks sane to me. I did not re-check anything content-wise.

_follow-up item up for discussion_:
Looking at the **Installing** Foreman Server guide again, I believe we should really split off "external auth" + "external services". Both chapters are not 100% related to purely installing Foreman. Now that we have assemblies it'll be a breeze.
image

**Comment by asteflova** [comment]
FYI, in the screenshot you posted Max, looking at Chapter 5., I consider only 5.1-5.8 to be user-story-based assemblies. 5.9-5.13 are rather just a loose collection of modules that don't really fit in any existing assembly. It's one of the things I want to improve with subsequent PRs for external authentication.

I'll let this PR open for a bit longer and merge tomorrow if there is no other feedback.

**Comment by maximiliankolb** [comment]
Yes, please merge Anet! The PR has my full support; I just wanted to hint at other topics you/we might want to tackle someday. :)

**Comment by ekohl** [comment]
> Looking at the **Installing** Foreman Server guide again, I believe we should really split off "external auth" + "external services". Both chapters are not 100% related to purely installing Foreman. Now that we have assemblies it'll be a breeze.

For what it's worth, @apinnick and I have been talking about that. More specifically, how to track the intent. Since it was brought up again, I've opened https://github.com/theforeman/foreman-documentation/issues/2957 as a dedicated way to track the effort.

I'd like to do the same with DHCP. It may make sense to start with that since there are fewer DHCP providers, making it easier. On the other hand, there's more to gain with DNS to shorten guides.

**Comment by lumarel** [comment]
> > Does this relate to what we discussed yesterday with @lumarel about the two Keycloak versions?
> 
> I don't think so, but then again, I have no idea who @lumarel is (hello!).

@asteflova Hi Aneta, I only now remembered that I did answer here 🙂 
Me just being here and there and trying to be of use, just a community member with "too much free time" ^^

**Pull request merged**